policyeval/policyserver.go

```diff
@@ -0,0 +1,142 @@
+package policyeval
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"maunium.net/go/mautrix/event"
+
+	"github.com/rs/zerolog"
+	"maunium.net/go/mautrix/federation"
+	"maunium.net/go/mautrix/id"
+
+	"go.mau.fi/meowlnir/policylist"
+)
+
+type psCacheEntry struct {
+	Recommendation PSRecommendation
+	Timestamp      time.Time
+	RoomID         id.RoomID
+	PDU            *event.Event
+}
+
+type PolicyServer struct {
+	Federation *federation.Client
+	ServerAuth *federation.ServerAuth
+	EventCache map[id.EventID]*psCacheEntry
+	cacheLock  *sync.RWMutex
+}
+
+func NewPolicyServer(serverName string) *PolicyServer {
+	inMemCache := federation.NewInMemoryCache()
+	fed := federation.NewClient(serverName, nil, inMemCache)
+	return &PolicyServer{
+		EventCache: make(map[id.EventID]*psCacheEntry),
+		cacheLock:  &sync.RWMutex{},
+		Federation: fed,
+		ServerAuth: federation.NewServerAuth(fed, inMemCache, func(auth federation.XMatrixAuth) string {
+			return auth.Destination
+		}),
+	}
+}
+
+func (ps *PolicyServer) getCachedRecommendation(evtID id.EventID) (*PolicyServerResponse, bool) {
+	ps.cacheLock.RLock()
+	resp, ok := ps.EventCache[evtID]
+	ps.cacheLock.RUnlock()
+	if ok {
+		if time.Since(resp.Timestamp) < 5*time.Minute {
+			return &PolicyServerResponse{Recommendation: resp.Recommendation}, true
+		}
+		ps.cacheLock.Lock()
+		delete(ps.EventCache, evtID)
+		ps.cacheLock.Unlock()
+	}
+	return nil, false
+}
+
+func (ps *PolicyServer) cacheRecommendation(evtID id.EventID, entry *psCacheEntry) {
+	ps.cacheLock.Lock()
+	defer ps.cacheLock.Unlock()
+	ps.EventCache[evtID] = entry
+	if len(ps.EventCache) > 1000 {
+		// clear out old entries to save space
+		for k := range ps.EventCache {
+			if time.Since(ps.EventCache[k].Timestamp) > 5*time.Minute {
```

Comment:
The size and age values should be extracted into constants to be neater.

It's also not optimal that 3+ events per second would lead to iterating over the map on every event, but it's not the end of the world